book/invalidation.md

```diff
@@ -2539,54 +2597,41 @@ I've left exploring it to an advanced exercise.
-that happened to be more overloaded than other ones. This led to the
-HTML parser for the certain test cases yielding[^parser-yield] more often,
-triggering different and incorrect rendering paths.
-
-Real browsers also make very aggressive use of assertions like those
-I added in this chapter. To avoid slowing down the browser for users,
-non-essential assertions are "compiled out" in what's usually called
-the *release build*, but left in for the *debug build*. The debug build
-is often used by engineers when debugging or developing new features,
-and also used to run automated tests.
+Real browsers also use assertions to catch bugs, much like the
+`ProtectedField` abstraction in this chapter. But to avoid slowing
+down the browser for users, non-essential assertions are "compiled
+out" in the *release build*, which is what end-users run. The *debug
+build*, which browser engineers use when debugging or developing new
+features. Automated tests also keep these assertions on. Debug builds
+also compile in debugging features like [sanitizers][ffx-sanitizers],
+while release builds instead use optimizations [like PGO][chrome-pgo].
 
 :::
 
-[^parser-yield]: In a real browser, HTML parsing doesn't happen in one go,
-but often is broken up into multiple event loop tasks. This leads to better
-web page loading performance, and is the reason you'll often see web pages
-render only part of the HTML at first when loading large web pages (including
-this book!).
+[ffx-sanitizers]: https://firefox-source-docs.mozilla.org/tools/sanitizer/index.html
+[chrome-pgo]: https://blog.chromium.org/2020/08/chrome-just-got-faster-with-profile.html
 
 Summary
 =======
 
 This chapter introduces the concept of partial style and layout through the
 technique of optimized cache invalidation. The main takeaways are:
 
-* Partial rendering dramatically speeds up key browser
-interactions, and are therefore an essential feature of real browsers.
+- Caching and invalidation is a powerful way to speeds up key browser
+  interactions, and is therefore an essential technique in real browsers.
 
-* Making rendering idempotent allows us to systematically remove work
- without worrying about side-effects.
+- Making rendering idempotent allows us skip redundant work
+  while guaranteeing that the page will look the same.
 
-* A good browser aims to preserve the principle of incremental
-performance---the cost of an update should be proportional to the change made.
+- A good browser aims for the principle of incremental performance:
+  the cost of an change should be proportional to size of the change,
+  not the size of the page as a whole.
 
-* Cache invalidation is difficult and error-prone, and justifies careful
-work and good abstractions like `ProtectedField`.
+- Cache invalidation is difficult and error-prone,
+  and justifies careful abstractions like `ProtectedField`.
+
+- Both data dependencies and control dependencies can be invalidated
```

Comment:
"control dependencies" were not introduced as a term of art in the latest draft of this chapter. Maybe rephrase to "data dependencies and tree walk dependencies"?

Reply:
Rewrote to "Invalidation can be used to skip allocation, computation, or even traversals."